byron/ledger/executable-spec/test/Generator.hs

```diff
@@ -0,0 +1,109 @@
+module Generator where
+
+import           Data.Bits (shiftL)
+import           Data.Word (Word64)
+import           Hedgehog
+import qualified Hedgehog.Gen    as Gen
+import qualified Hedgehog.Range  as Range
+import           Ledger.Core (Slot(..), SlotCount(..), BlockCount(..))
+import           Ledger.GlobalParams (k)
+import           Ledger.Update (ProtVer(..), PParams(..))
+import           Numeric.Natural (Natural)
+
+-- | Generates a slot within the given bound
+slotGen :: Word64 -> Word64 -> Gen Slot
+slotGen lower upper = do
+  Slot <$> Gen.word64 (Range.linear lower upper)
+
+-- | Generates a protocol version
+protVerGen :: Gen ProtVer
+protVerGen = do
+  pvMin  <- Gen.integral (Range.linear (0 :: Natural) 100)
+  pvDiff <- Gen.integral (Range.linear (0 :: Natural) 20)
+  let pvMax = pvMin + pvDiff
+  pvAlt  <- Gen.integral (Range.linear (0 :: Natural) 100)
+  return $ ProtVer pvMin pvMax pvAlt
+
+-- | Generates a Double in the inclusive range [0, 1]
+doubleGen :: Gen Double
+doubleGen = do
+  b <- Gen.word64 (Range.linear 1 hi)
+  a <- Gen.word64 (Range.linear 0 b )
+  return $ (fromIntegral a) / (fromIntegral b)
+ where
+  hi = 2 `shiftL` 48 :: Word64
+
+-- | Generates valid protocol parameters
+pparamsGen :: Gen PParams
+pparamsGen = do
+  maxBkSz         <- Gen.integral (Range.linear 1 1000)
+  maxHdrSz        <- Gen.integral (Range.linear 1 maxBkSz)
+  maxTxSz         <- Gen.integral (Range.linear 1 maxBkSz)
+  maxPropSz       <- Gen.integral (Range.linear 1 maxBkSz)
+  bkSgnCntT       <- doubleGen
+  bkSlotsPerEpoch <- SlotCount <$> Gen.word64 (Range.linear 1 21600)
+  upTtl           <- SlotCount <$> Gen.word64
+    (Range.linear 1 (unSlotCount bkSlotsPerEpoch))
+  scriptVersion   <- Gen.integral (Range.linear (0 :: Natural) 1000)
+  cfmThd          <- Gen.integral (Range.linear 0 1000)
+  upAdptThd       <- Gen.integral (Range.linear 1 100)
+  stableAfter     <- BlockCount <$> Gen.word64
+    (Range.linear 1 (unSlotCount bkSlotsPerEpoch))
+  return $
+    PParams
+      maxBkSz
+      maxHdrSz
+      maxTxSz
+      maxPropSz
+      bkSgnCntT
+      bkSlotsPerEpoch
+      upTtl
+      scriptVersion
+      cfmThd
+      upAdptThd
+      stableAfter
+
+data PVBUMPEnv = PVBUMPEnv
+  ( Slot
+  , [(Slot, (ProtVer, PParams))]
+  ) deriving Show
+
+-- | Generates an environment for the PVBUMP STS
+pvbumpEnvGen :: Gen PVBUMPEnv
```

Comment:
So to clarify what I said above, regardless of what changes we make in the future, here we could simply write:

```haskell
pvbumpEnvGen :: Gen (Environment PVBUMP)

```

Reply:
So no need for new `data` or `type` aliases.

Reply:
I use a slightly different approach here, as explained earlier.

Reply:
I implemented what you suggested.